### PR TITLE
Honor `output_type=tf.int32` in `Argmax` / `Argmin` generators

### PR DESCRIPTION
**Pull request by khatchad**
## Summary

Fixes [wala/ML#463](https://github.com/wala/ML/issues/463).

`tf.math.argmax` / `tf.math.argmin` accept an `output_type` keyword that overrides the default `int64` index dtype with `int32` (the only other accepted value per the TF 2.9 docs). The XML modeling and the `Argmax` Java generator both ignored this argument: the XML didn't include `output_type` in `paramNames`, so the analyzer couldn't even resolve the keyword; and `Argmax.getDTypes` returned `INT64` unconditionally regardless.

## Changes

- **XML** (`tensorflow.xml`): `argmax` and `argmin` blocks now expose `output_type` in `paramNames` (`numArgs=6`, position matching the TF 2.x signature `(input, axis, output_type, name)` — trailing `dimension` retained as the TF 1.x alias for `axis`).
- **Java** (`Argmax.java`):
  - Adds a `Parameters` enum mirroring the new XML positions.
  - Overrides `getDTypeParameterPosition` / `getDTypeParameterName` to point at `output_type`.
  - Replaces the `getDTypes` override that returned `INT64` unconditionally with `getDefaultDTypes` returning `INT64`; a targeted `getDTypes` override inlines the canonical `TensorGenerator.getDTypes` dispatch (read dtype-arg if present; fall back to `getDefaultDTypes`) — bypassing `ReduceMean`'s inherit-from-input dtype path, which is wrong for argmax (the output is an integer index regardless of input dtype).
  - `Argmin` inherits the new behavior unchanged via class extension.
- **Tests**: two new fixtures + JUnit tests (`testArgmaxOutputType`, `testArgminOutputType`) asserting the `int32` override; new `TENSOR_INT32_UNKNOWN_SHAPE` constant.

## Test Plan

- [x] Full ML test suite passes locally: `658 tests, 0 failures, 0 errors, 0 skipped`.
- [x] Existing `testArgmax` / `testArgmin` continue to pass — the new XML position for `output_type` doesn't break the absent-`output_type` defaults path.
- [x] New `testArgmaxOutputType` / `testArgminOutputType` pass.
- [x] Both Python fixtures run to completion under `python3.10` with the documented `tf.int32` runtime assertion.
- [ ] CI confirms.

## Why The `getDTypes` Override Inlines The Base Logic

`Argmax extends ReduceMean extends TensorGenerator`. `ReduceMean.getDTypes` is overridden to return the input tensor's dtype — correct for `tf.reduce_mean`, wrong for argmax (which always returns an integer regardless of input). To use the canonical `TensorGenerator.getDTypes` dispatch (which checks the dtype arg, then falls back to `getDefaultDTypes`), Argmax has to bypass `ReduceMean`'s override; Java doesn't allow `super.super.method()`, so the canonical logic is inlined in `Argmax.getDTypes` (a four-line read-arg-or-fall-back-to-default body). If the `Argmax` / `ReduceMean` hierarchy is ever revisited (e.g. via a unified-dispatch refactor under wala/ML#469), this duplication can be eliminated.

## Related

- [wala/ML#463](https://github.com/wala/ML/issues/463) — this PR's tracking issue.
- [wala/ML#449](https://github.com/wala/ML/issues/449) — Tier roadmap (this is part of the Tier 6 cleanup).
- [wala/ML#462](https://github.com/wala/ML/issues/462) — output-shape precision is still deferred (separate issue, not addressed here).
- [wala/ML#472](https://github.com/wala/ML/issues/472) — design discussion on `PassThroughUnaryTensorGenerator`'s fixed-dtype subclasses; this commit's pattern (override `getDefaultDTypes` to return a constant + override `getDType*` getters to expose the override arg) lines up with what `SequenceMask` would do once `dtype` is wired through.

### CODE REVIEW

File: com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# argmax output dtype is `int64`; passing `output_type=tf.int32` should
+# override it to `int32`. The static analysis reads `output_type` from
+# `tensorflow.xml`'s `paramNames` and routes through the inherited
+# dtype-arg machinery.
+x = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+y = tf.math.argmax(x, axis=0, output_type=tf.int32)
+assert isinstance(y, tf.Tensor)
+assert y.dtype == tf.int32
+
+f(y)
```
Comment (khatchad) [inline com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py]: Add `x` here also.

File: com.ibm.wala.cast.python.test/data/tf2_test_argmin_output_type.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+
+# Counterpart of `tf2_test_argmax_output_type.py` for `tf.math.argmin`.
+# `Argmin` extends `Argmax`, so the same `output_type` machinery applies.
+x = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+y = tf.math.argmin(x, axis=0, output_type=tf.int32)
+assert isinstance(y, tf.Tensor)
+assert y.dtype == tf.int32
+
+f(y)
```
Comment (khatchad) [inline com.ibm.wala.cast.python.test/data/tf2_test_argmin_output_type.py]: Add `x` here also.

File: com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+# Drives the `output_type=tf.int32` branch of `Argmax.getDTypes`. Default
+# argmax output dtype is `int64`; passing `output_type=tf.int32` should
+# override it to `int32`. The static analysis reads `output_type` from
+# `tensorflow.xml`'s `paramNames` and routes through the inherited
+# dtype-arg machinery.
+x = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+y = tf.math.argmax(x, axis=0, output_type=tf.int32)
+assert isinstance(y, tf.Tensor)
+assert y.dtype == tf.int32
```
Comment (khatchad) [inline com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py:19]: Why no `shape` assertion for `y`?

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 4]
```diff
@@ -340,6 +340,8 @@ public class TestTensorflow2Model extends TestPythonMLCallGraphShape {
 
   private static final TensorType TENSOR_INT64_UNKNOWN_SHAPE = new TensorType(INT_64, null);
 
+  private static final TensorType TENSOR_INT32_UNKNOWN_SHAPE = new TensorType(INT_32, null);
```
Comment (khatchad) [inline com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:343]: Why is the shape unknown?

File: com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py (hunk 1 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py]: Done in a46af15c. Switched the sink to two parameters (`def f(x, y): pass; f(x, y)`) so each tensor's type is checked independently &mdash; parameter `2` is `x`'s `(2, 3) float32`, parameter `3` is `y`'s `int32` (with ⊤ shape, see thread on `TENSOR_INT32_UNKNOWN_SHAPE`).

I also added a companion `testArgmaxOutputTypeDoubleSink` (and `Argmin` counterpart) that exercises the *single-parameter sink, two call sites* shape (`def f(a): pass; f(x); f(y)`) &mdash; this is the union-across-call-sites case you wondered about. It produces `{(2, 3) float32, ? int32}` precisely, confirming the union holds and the `output_type=tf.int32` override survives both call sites. (Earlier I thought I'd seen this regress to `int64`; turned out to be a stale `~/.m2` snapshot from a different branch &mdash; cleared with `mvn clean install` and the precise answer reproduces. Filed and then closed wala/ML#490 with that explanation.)

File: com.ibm.wala.cast.python.test/data/tf2_test_argmin_output_type.py (hunk 2 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.test/data/tf2_test_argmin_output_type.py]: Done in a46af15c. Switched the sink to two parameters (`def f(x, y): pass; f(x, y)`) so each tensor's type is checked independently &mdash; parameter `2` is `x`'s `(2, 3) float32`, parameter `3` is `y`'s `int32` (with ⊤ shape, see thread on `TENSOR_INT32_UNKNOWN_SHAPE`).

I also added a companion `testArgmaxOutputTypeDoubleSink` (and `Argmin` counterpart) that exercises the *single-parameter sink, two call sites* shape (`def f(a): pass; f(x); f(y)`) &mdash; this is the union-across-call-sites case you wondered about. It produces `{(2, 3) float32, ? int32}` precisely, confirming the union holds and the `output_type=tf.int32` override survives both call sites. (Earlier I thought I'd seen this regress to `int64`; turned out to be a stale `~/.m2` snapshot from a different branch &mdash; cleared with `mvn clean install` and the precise answer reproduces. Filed and then closed wala/ML#490 with that explanation.)

File: com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py (hunk 3 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py:19]: Added in a46af15c. The fixture now asserts `x.shape == (2, 3)`, `x.dtype == tf.float32`, `y.shape == (3,)` (and `y.dtype == tf.int32` was already there). I re-ran `python3.10 com.ibm.wala.cast.python.test/data/tf2_test_argmax_output_type.py` and `..._argmin_..py` to confirm: both pass with no traceback, so the runtime answer matches the JUnit expectation on every dimension we can express today.

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java (hunk 4 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:343]: The shape on `y` is currently ⊤ because `Argmax` deliberately doesn't compose `input.shape[:axis] + input.shape[axis+1:]` &mdash; the class-level Javadoc spells out why:

> Output shape is left at ⊤ for now; computing the precise shape (input.shape with the `axis` dimension removed) regresses tests such as `testNeuralNetwork*`, where the per-context shape union for `y_true` (e.g. `[256]` train vs. `[10000]` test) cross-products through `ElementWiseOperation`'s strict broadcast check.

So the gap is gated by wala/ML#462 (`ElementWiseOperation` cartesian-pair throw defeats per-context shape precision in generators with concrete shape inference). I added a docstring + TODO on `testArgmaxOutputType` in a46af15c that names #462 explicitly so the tightening is discoverable when #462 lands.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 5]
```diff
@@ -2,27 +2,77 @@
...
+
+    /**
+     * Lowercase keyword name used in argument-resolution helpers when the call site uses {@code
+     * keyword=value} syntax.
+     *
+     * @return The lowercased enum name (e.g. {@code "output_type"}).
+     */
+    public String getName() {
+      return name().toLowerCase();
+    }
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java:70]: Fixed in fbe44700 &mdash; switched `Argmax.Parameters.getName()` to `name().toLowerCase(Locale.ROOT)` so the conversion is locale-stable. Filed wala/ML#497 to sweep the same fix across the other `Parameters` enums I've added in this generator batch (`Where`, `Einsum`, `Meshgrid`, `LeakyRelu`, the 17 tier-1 unary generators on PR ponder-lab/ML#258, etc.) so the locale-stability is consistent.

File: com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java [hunk 6]
```diff
@@ -2,27 +2,77 @@
...
- * default; the {@code output_type=tf.int32} alternative isn't currently surfaced through {@code
- * tensorflow.xml}, so the generator emits {@code int64} unconditionally). Output shape is left at ⊤
- * for now; computing the precise shape (input.shape with the {@code axis} dimension removed)
- * regresses tests such as {@code testNeuralNetwork*}, where the per-context shape union for {@code
- * y_true} (e.g. {@code [256]} train vs. {@code [10000]} test) cross-products through {@code
+ * Generator for {@code tf.math.argmax}. Output dtype defaults to {@link DType#INT64} (the TF
+ * default); when the call passes an explicit {@code output_type} argument (e.g. {@code
+ * output_type=tf.int32}), the inherited dtype-arg machinery in {@link TensorGenerator#getDTypes}
+ * resolves it via {@link #getDTypeParameterPosition} / {@link #getDTypeParameterName} and uses the
+ * resolved dtype instead of the default — fix for <a
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml/source/com/ibm/wala/cast/python/ml/client/Argmax.java]: Fixed in fbe44700. The class-level Javadoc lead now states the dispatch is *inlined* in `Argmax.getDTypes` (not inherited), and explains why: bypassing `ReduceMean#getDTypes` is necessary because argmax always returns an integer index regardless of input dtype.

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 7]
```diff
@@ -4507,6 +4509,119 @@ public void testArgmin()
...
+  public void testArgmaxOutputType()
+      throws ClassHierarchyException, IllegalArgumentException, CancelException, IOException {
+    test(
+        "tf2_test_argmax_output_type.py",
+        "f",
+        2,
+        2,
+        Map.of(
+            2, Set.of(TENSOR_2_3_FLOAT32),
+            3, Set.of(TENSOR_INT32_UNKNOWN_SHAPE)));
```
Comment (khatchad) [inline com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:4550]: I see an unknown shape here but no TODO comment.

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java (hunk 4 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:343]: Superceded by https://github.com/ponder-lab/ML/pull/246/changes#diff-766d90d7068cc279c23eef8479fc79346a535181a799e9437d93e00806f02d91R4540.

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java (hunk 7 above)
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java:4550]: > I see an unknown shape here but no TODO comment.

Right &mdash; the Javadoc explained the imprecision and referenced wala/ML#462 but didn't use the explicit `TODO(<issue>):` marker that the prefer-observed-assertion convention calls for. Converted to that format in 39fc3faf, and while I was at it, added matching markers to the sibling tests `testArgmax` and `testArgmin` &mdash; both had the same ⊤ shape imprecision and prose-only Javadoc.

- `testArgmaxOutputType`: https://github.com/ponder-lab/ML/blob/39fc3fafe47c25518eddae7a0fd6fc7ec439939c/com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java#L4519-L4523
- `testArgmax` / `testArgmin`: https://github.com/ponder-lab/ML/blob/39fc3fafe47c25518eddae7a0fd6fc7ec439939c/com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java#L4480-L4517

File: com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java [hunk 8]
```diff
@@ -4496,6 +4503,10 @@ public void testArgmax()
    * Counterpart of {@link #testArgmax()} for {@code tf.math.argmin}. Same semantics: dtype is fixed
    * at {@code int64}, shape is left at ⊤. See {@link com.ibm.wala.cast.python.ml.client.Argmin}.
```
[the remark this answers is not in the document]
Comment (khatchad) [reply com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java]: > The `testArgmin` Javadoc says argmin's dtype is "fixed at int64", but this PR adds `output_type` support (see `testArgminOutputType` below) so the dtype is only *defaulted* to `int64` when `output_type` is absent. Update the wording to avoid documenting incorrect semantics for `tf.math.argmin`.

Right &mdash; "fixed at" was leftover from before the `output_type` work. Reworded to "defaults to int64 (overridable via output_type)" in cd997186, with a cross-reference to `testArgminOutputType`: https://github.com/ponder-lab/ML/blob/cd997186c8034e5e8e4220144af31b0bc40346e7/com.ibm.wala.cast.python.ml.test/source/com/ibm/wala/cast/python/ml/test/TestTensorflow2Model.java#L4502-L4505.
